### PR TITLE
Refactor layer property handling

### PR DESCRIPTION
**Pull request by ibgreen**
Lays the foundations for the next PR  implementing "async props" ("prop rewriting") for dataUrl RFC.

Note: This will be merged into a 6.0-dev branch, no risk to master stability. Review is of course still encouraged.

### CODE REVIEW

File: src/core/lib/layer-props.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    this._layer = layer;
+    props = this.normalize(props);
+    Object.assign(this, props);
+    // Props are immutable
+    Object.freeze(this);
+  }
+
+  normalize(props) {
+    // If sublayer has static defaultProps member, getDefaultProps will return it
+    const mergedDefaultProps = getDefaultProps(this._layer);
```
Comment (Pessimistress) [inline src/core/lib/layer-props.js]: Layers are constructed frequently but `mergedDefaultProps` doesn't change for each class. Might help perf if we save it somewhere?

File: src/core/lib/layer.js [hunk 2]
```diff
@@ -749,11 +725,11 @@ ${flags.viewportChanged ? 'viewport' : ''}\
...
-    this.state.needsRedraw = true;
+    this.setNeedsRedraw();
     log.deprecated('layer.setUniforms', 'model.setUniforms');
   }
 }
 
 Layer.layerName = 'Layer';
-Layer.propTypes = defaultProps;
-Layer.defaultProps = defaultProps;
+Layer.propTypes = LayerProps.defaultProps;
```
Comment (Pessimistress) [inline src/core/lib/layer.js]: `propTypes`?

File: src/core/lib/layer-props.js (hunk 1 above)
Comment (ibgreen) [reply src/core/lib/layer-props.js]: > Layers are constructed frequently but mergedDefaultProps doesn't change for each class. Might help perf if we save it somewhere?

We do save it [inside `getDefaultProps`](https://github.com/uber/deck.gl/blob/master/src/core/lib/props.js#L178) on the constructor' prototype. However for some reason accessing the prototype is very slow, we should make a separate cache instead. 

But unrelated to this PR.

File: src/core/lib/layer.js (hunk 2 above)
Comment (ibgreen) [reply src/core/lib/layer.js]: > propTypes?
Good catch, missed removing this when splitting the diff...

**Comment by ibgreen** [comment]
This will be merged into a 6.0-dev branch. Review is of course still encouraged.

**Pull request merged**